core/shared/src/main/scala/scalaz/zio/Queue.scala

```diff
@@ -3,156 +3,406 @@
     }
 
     val release: (Boolean, Promise[Nothing, A]) => IO[Nothing, Unit] = {
-      case (_, p) => removeTaker(p)
+      case (_, p) => p.poll.void <> removeTaker(p)
     }
+
     Promise.bracket[Nothing, State[A], A, Boolean](ref)(acquire)(release)
+
   }
 
   /**
-   * Interrupts any fibers that are suspended on `take` because the queue is
-   * empty. If any fibers are interrupted, returns true, otherwise, returns
-   * false.
+   * Takes up to max number of values in the queue.
    */
-  final def interruptTake: IO[Nothing, Boolean] =
-    IO.flatten(ref.modify {
-      case Deficit(takers) if takers.nonEmpty =>
-        val forked: IO[Nothing, Fiber[Nothing, List[Boolean]]] =
-          IO.forkAll[Nothing, Boolean](takers.map(_.interrupt))
-        (forked.flatMap(_.join).map(_.forall(identity)), Deficit(IQueue.empty[Promise[Nothing, A]]))
-      case s =>
-        (IO.now(false), s)
+  final def takeUpTo(max: Int): IO[Nothing, List[A]] =
+    IO.flatten(ref.modify[IO[Nothing, List[A]]] {
+      case Surplus(values, putters, hook) =>
+        val (q1, q2)             = values.splitAt(max)
+        val (newState, promises) = moveNPutters(Surplus(q2, putters, hook), q1.size)
+        (promises *> IO.point(q1.toList), newState)
+      case state @ Deficit(_, _) => (IO.now(Nil), state)
+      case state @ Shutdown      => (IO.interrupt, state)
     })
 
   /**
-   * Interrupts any fibers that are suspended on `offer` because the queue is
-   * at capacity. If any fibers are interrupted, returns true, otherwise,
-   * returns  false.
+   * Interrupts any fibers that are suspended on `offer` or `take`.
+   * Future calls to `offer*` and `take*` will be interrupted immediately.
    */
-  final def interruptOffer: IO[Nothing, Boolean] =
+  final val shutdown: IO[Nothing, Unit] =
     IO.flatten(ref.modify {
-      case Surplus(_, putters) if putters.nonEmpty =>
-        val forked: IO[Nothing, Fiber[Nothing, List[Boolean]]] =
-          IO.forkAll[Nothing, Boolean](putters.map(_._2.interrupt))
-        (forked.flatMap(_.join).map(_.forall(identity)), Deficit(IQueue.empty[Promise[Nothing, A]]))
-      case s =>
-        (IO.now(false), s)
+      case Surplus(_, putters, hook) if putters.nonEmpty =>
+        val forked = IO
+          .forkAll[Nothing, Boolean](putters.toList.map {
+            case (_, p) => p.interrupt
+          })
+          .flatMap(_.join)
+        (forked *> hook, Shutdown)
+      case Deficit(takers, hook) if takers.nonEmpty =>
+        val forked = IO
+          .forkAll[Nothing, Boolean](
+            takers.toList.map(p => p.interrupt)
+          )
+          .flatMap(_.join)
+        (forked *> hook, Shutdown)
+      case Shutdown => (IO.unit, Shutdown)
+      case state    => (state.shutdownHook, Shutdown)
     })
 
-  private final def removePutter(putter: Promise[Nothing, Unit]): IO[Nothing, Unit] =
+  final private def removePutter(putter: Promise[Nothing, Boolean]): IO[Nothing, Unit] =
```

Comment:
For organization purposes, can we:

1. Put all the public methods at the front of the class, maybe in alphabetical order.
2. Put all the private methods at the end.

This way it's easier for users who are browsing the code to see the most important pieces.